src/libproc/proc_pid.rs

```diff
@@ -637,3 +637,452 @@ impl From<uint32_t> for ProcFDType {
     }
 }
 
+// This trait is needed for polymorphism on pidfdinfo types, also abstracting flavor in order to provide
+// type-guaranteed flavor correctness
+pub trait PIDFDInfo: Default {
+    fn flavor() -> PidFDInfoFlavor;
+}
+
+/// Returns the information about file descriptors of the process that match pid passed in.
+///
+/// # Examples
+///
+/// ```
+/// use std::io::Write;
+/// use std::net::TcpListener;
+/// use libproc::libproc::proc_pid::{listpidinfo, pidinfo, pidfdinfo, ListFDs, ListThreads, BSDInfo, ProcFDType, SocketFDInfo, SocketInfoKind};
+///
+/// fn pidfdinfo_test() {
+///     use std::process;
+///     let pid = process::id() as i32;
+///
+///     // Open TCP port:8000 to test.
+///     let _listener = TcpListener::bind("127.0.0.1:8000");
+///
+///     if let Ok(info) = pidinfo::<BSDInfo>(pid, 0) {
+///         if let Ok(fds) = listpidinfo::<ListFDs>(pid, info.pbi_nfiles as usize) {
+///             for fd in &fds {
+///                 match fd.proc_fdtype.into() {
+///                     ProcFDType::Socket => {
+///                         if let Ok(socket) = pidfdinfo::<SocketFDInfo>(pid, fd.proc_fd) {
+///                             match socket.psi.soi_kind.into() {
```

Comment:
You could reduce indentation one layer here :-) by matching Ok(SocketInfo...) if you wanted....